zokrates_cli/src/ops/mpc/beacon.rs

```diff
@@ -0,0 +1,152 @@
+
+    let reader = BufReader::new(file);
+    let mut params = MPCParameters::read(reader, true)
+        .map_err(|why| format!("Could not read `{}`: {}", path.display(), why))?;
+
+    let beacon_hash = sub_matches.value_of("hash").unwrap();
+    let num_iterations: usize = sub_matches.value_of("iterations").unwrap().parse().unwrap();
+
+    if !(10..=63).contains(&num_iterations) {
+        return Err("Number of hash iterations should be in [10, 63] range".to_string());
+    }
+
+    println!("Creating a beacon RNG");
+
+    // Create an RNG based on the outcome of the random beacon
+    let mut rng = {
+        use byteorder::{BigEndian, ReadBytesExt};
+        use crypto::digest::Digest;
+        use crypto::sha2::Sha256;
+        use rand::chacha::ChaChaRng;
+        use rand::SeedableRng;
+
+        // The hash used for the beacon
+        let hash_result = hex::decode(beacon_hash);
+        if hash_result.is_err() {
+            return Err("Beacon hash should be in hexadecimal format".to_string());
+        }
+
+        let mut cur_hash = hash_result.unwrap();
+        if cur_hash.len() != 32 {
+            return Err("Beacon hash should be 32 bytes long".to_string());
+        }
+
+        // Performs 2^n hash iterations over it
+        let n: usize = num_iterations;
+
+        for i in 0..(1u64 << n) {
+            // Print 1024 of the interstitial states
+            // so that verification can be
+            // parallelized
+            if i % (1u64 << (n - 10)) == 0 {
+                print!("{}: ", i);
+                for b in cur_hash.iter() {
+                    print!("{:02x}", b);
+                }
+                println!();
+            }
+
+            let mut h = Sha256::new();
+            h.input(&cur_hash);
+            h.result(&mut cur_hash);
+        }
+
+        print!("Final result of beacon: ");
+        for b in cur_hash.iter() {
+            print!("{:02x}", b);
+        }
+        println!("\n");
+
+        let mut digest = &cur_hash[..];
+
+        let mut seed = [0u32; 8];
+        for e in &mut seed {
+            *e = digest.read_u32::<BigEndian>().unwrap();
+        }
+
+        ChaChaRng::from_seed(&seed)
+    };
+
+    println!("Contributing to `{}`...", path.display());
+    let zero: u32 = 0;
+    let hash = params.contribute(&mut rng, &zero);
+
+    println!("The BLAKE2b hash of your contribution is:\n");
+    for line in hash.chunks(16) {
+        print!("\t");
+        for section in line.chunks(4) {
+            for b in section {
+                print!("{:02x}", b);
+            }
+            print!(" ");
+        }
+        println!();
+    }
+
+    let output_path = Path::new(sub_matches.value_of("output").unwrap());
+    let output_file = File::create(&output_path)
+        .map_err(|why| format!("Could not create `{}`: {}", output_path.display(), why))?;
+
+    println!(
+        "\nYour contribution has been written to `{}`",
```

Comment:
Seems like this command is doing both contributing and applying the beacon. Is there a way to decorrelate the two? Or do I understand correctly that this is basically a normal contribution but using a specific deterministic randomness instead of some private randomness, so it almost sounds like a special case of contribution? Could this be exposed as a flag on the contribution command instead of this separate command? Not necessarily suggesting to do that, but conceptually?

Reply:
From what I understand applying the beacon is actually the same as contributing, the only difference is how the rng is created. I've separated it because that is the flow in other available tools, so we don't differentiate that much.